clang/test/SemaSYCL/num_simd_work_items_device.cpp

```diff
@@ -89,15 +89,15 @@ struct TRIFuncObjBad7 {
 };
 
 struct TRIFuncObjBad8 {
-  [[intel::num_simd_work_items(3)]]     // expected-error{{'num_simd_work_items' attribute must evenly divide the work-group size for the 'reqd_work_group_size' attribute}}
-  [[intel::reqd_work_group_size(5, 5)]] // expected-note{{conflicting attribute is here}}
+  [[intel::num_simd_work_items(4)]]      // expected-error{{'num_simd_work_items' attribute must evenly divide the work-group size for the 'reqd_work_group_size' attribute}}
```

Comment:
I think I've misunderstood the bug and fix. I thought for SYCL spelling, the last argument needs to be evenly divisible. That understanding goes against this test. Can you explain required behavior please?

Reply:
Thank you @elizabethandrews for taking a look into the patch. OpenCL and SYCL design are different. The order of the arguments to reqd_work_group_size in SYCL and OpenCL are reversed. In the OpenCL syntax, reqd_work_group_size(X,Y,Z), X is the index that increments the fastest, followed by Y, then Z. If we preserve that definition of X, Y, and Z, then in SYCL the attribute is specified as reqd_work_group_size(Z,Y,X).  So X (fist argument) goes to last, This is what i am doing here. I am checking whether this is divided by num_simd attribute.

Reply:
Do you mean (X, Y, Z) for SYCL or (Z, Y, X) for SYCL?   Because in your tests, you are checking against the first argument.

Reply:
Based on the discussion here : https://github.com/intel/llvm/issues/3252, my understanding was that X argument (first argument) of reqd_work_group_size attribute need to be checked here whether it is divisible by num_simd_work_items attribute or not. Tagging @AaronBallman here.

As per compiler source codes, we do not swap the reqd_work_group_size attribute values anymore during semantic checking but we do for codegen part on https://github.com/intel/llvm/pull/2962.

struct TRIFuncObjGood8 {
  [[intel::reqd_work_group_size(64, 5)]]
  [[intel::num_simd_work_items(4)]] void
  operator()() const {}
};

AST:

 h.single_task<class test_kernel11>(TRIFuncObjGood8());
    // CHECK-LABEL: FunctionDecl {{.*}}test_kernel11
    // CHECK:       ReqdWorkGroupSizeAttr {{.*}}
    // CHECK-NEXT:  ConstantExpr{{.*}}'int'
    // CHECK-NEXT:  value: Int 64
    // CHECK-NEXT:  IntegerLiteral{{.*}}64{{$}}
    // CHECK-NEXT:  ConstantExpr{{.*}}'int'
    // CHECK-NEXT:  value: Int 5
    // CHECK-NEXT:  IntegerLiteral{{.*}}5{{$}}
    // CHECK-NEXT:  ConstantExpr{{.*}}'int'
    // CHECK-NEXT:  value: Int 1
    // CHECK-NEXT:  IntegerLiteral{{.*}}1{{$}}
    // CHECK:       SYCLIntelNumSimdWorkItemsAttr {{.*}}
    // CHECK-NEXT:  ConstantExpr{{.*}}'int'
    // CHECK-NEXT:  value: Int 4
    // CHECK-NEXT:  IntegerLiteral{{.*}}4{{$}}

I meant to say (X, Y, Z) for SYCL, i am checking X argument (first) here.